lucille-core/src/main/java/com/kmwllc/lucille/stage/Chunking.java

```diff
@@ -0,0 +1,247 @@
+    }
+    if (characterLimit < -1) {
+      throw new StageException("Character limit must be a positive integer");
+    }
+    if (text_field == null) {
+      throw new StageException("text_field configuration where Chunk Stage would take as input is missing");
+    }
+    if (method == ChunkingMethod.CUSTOM && separator.isEmpty()) {
+      throw new StageException("Provide a non empty regEx for separator configuration");
+    }
+    if (method == ChunkingMethod.FIXED && characterLimit <= 0) {
+      throw new StageException("Provide a positive character_limit for fixed sized chunking");
+    }
+  }
+
+  @Override
+  public void start() throws StageException {
+    // load sentence model if we are using that chunking method
+    if (method == ChunkingMethod.SENTENCE) {
+      try (InputStream sentModelIn = getClass().getResourceAsStream("/en-sent.bin")) {
+        if (sentModelIn == null) {
+          throw new StageException("No sentence model found");
+        }
+        SentenceModel sentModel = new SentenceModel(sentModelIn);
+        sentenceDetector = new SentenceDetectorME(sentModel);
+      } catch (IOException e) {
+        throw new StageException("Could not load sentence model", e);
+      }
+    }
+  }
+
+  @Override
+  public Iterator<Document> processDocument(Document doc) throws StageException {
+    // retrieve content to chunk
+    String content = doc.getString(text_field);
+    // for testing: log.info("content retrieved {} ", content);
+
+    // splitting up content based on chunking method
+    String[] chunks;
+    switch (method) {
+      case CUSTOM:
+      case PARAGRAPH:
+        // regex pattern for paragraph: find any consecutive line break sequence (\n, \r, \r\n) within one unit of whitespace
+        String regex = (method == ChunkingMethod.CUSTOM) ? separator : "\\s*(?>\\R)\\s*(?>\\R)\\s*";
+        chunks = content.split(regex);
+        break;
+      case FIXED:
+        // splitting by characterLimit fails to split properly when there is line break sequence (\n, \r, \r\n) in content,
+        // so have decided to remove them before
+        content = content.replaceAll("(?>\\R)", " ");
+        chunks = content.split("(?<=\\G.{" + characterLimit + "})");
+        break;
+      default: // SENTENCE
+        chunks = sentenceDetector.sentDetect(content);
+        break;
+    }
+
+    // fixed method has already replaced all newline characters with " " before splitting
+    if (method != ChunkingMethod.FIXED) {
+      // replacing all newline characters in chunks
+      for (int i = 0; i < chunks.length; i++) {
+        chunks[i] = chunks[i].replaceAll("(?>\\R)", " ");
+      }
+    }
+
+    // truncating if we have character limit set
+    if (method != ChunkingMethod.FIXED && characterLimit > 0) {
+      truncateRest(chunks, characterLimit);
+    }
+
+    // overlap chunks if we have buffer set
+    if (newChunkSize > 1) chunks = overlapChunks(chunks);
```

Comment:
In the case where character limit & overlap is set, we need to discuss if we should truncate before overlap, or after overlap.

Reply:
I think the character limit should be applied after the overlap. If you set a character limit of 50 characters, you would expect the resulting field to have no more than 50 characters. However, with overlap after truncate this does not always hold true.

Reply:
There could be cases where we want to merge 3 sentences together and it turns out that the second sentence unexpectedly exceeds the character limit, and the final merged chunk after truncating would wipe out the last sentence which might have provided key information/context. If I think about the purpose of character limit it would lean toward as a precaution setting to limit a chunk before merging, and then the concern of fitting into a token limit would be done in an embedding or another stage.

Reply:
After further discussion with Brian and Spencer, concluded limit would be better to be placed after. Will implement option to truncate before merging and overlap.